# 函数返回值

* 异步编程有很多特有的代码设计模式

## 函数返回值

使用一个函数的输出, 作为另一个函数的输入
* 注意, 过多的嵌套容易写出`>`形状的代码

```javascript
// 同步的方式
var output = func1(func2('input'));
```

异步方式下, 函数执行并没有返回结果, 而是需要通过回调函数传递

```javascript
// 异步的方式, 通过回调函数传递
func2('intput', function(output2) {
    func1(output2, function(output1) {
        // do something
    });
});
```

## 遍历数组

在遍历数组时, 使用某个函数一次对数组成员做一些处理

```javascript
// 同步的方式
var len = arr.length,
    i = 0;
for(; i , len; ++i) {
    arr[i] = sync(arr[i]);
}
```

异步方式下, 需要使用回调函数嵌套处理

```javascript
// 异步的方式
(function next(i, len, callback) {
    if(i < len) {
        async(arr[i], function(value) {
            arr[i] = value;
            next(i + 1, len, callback);
        });
    } else {
        callback();
    }
}(0, arr.length, function() {
    // 所有元素已经处理完毕
}));
```

如果数组元素可以并行处理, 但后续代码仍然需要所有数组成员处理完毕才能执行, 则可以调整代码如下

```javascript
(function (i, len, count, callback) {
    for (; i < len; ++i) {
        (function (i) {
            async(arr[i], function (value) {
                arr[i] = value;
                if (++count === len) {
                    callback();
                }
            });
        }(i));
    }
}(0, arr.length, 0, function () {
    // All array items have processed.
}));
```

## 异常处理

* `try...catch...`可以进行异常捕获
* 注意, 在异步函数中, 异常的捕获位置需要特殊注意: 要在异常冒泡到断点之前用`try...catch`捕获, 并通过回调函数传递被捕获的异常
* JS中, 异步API都是按照回调函数第一个参数为`err`设计的

```javascript
// 同步异常
function sync(fn) {
    return fn();
}

try {
    sync(null);
    // do something
} catch (err) {
    console.log('Error: %s', err.message);
}
```

执行脚本

```shell
$ node tryCatch.js
Error: fn is not a function
```

```javascript
// 异步函数的异常捕获
function async(fn, callback) {
    // Code execution path breaks here.
    setTimeout(function ()　{
        try {
            callback(null, fn());
        } catch (err) {
            callback(err);
        }
    }, 0);
}

async(null, function (err, data) {
    if (err) {
        console.log('Error: %s', err.message);
    } else {
        // Do something.
    }
});
```

执行脚本

```shell
$ node tryCatchAsync.js
Error: fn is not a function
```


